src/Imgix/UrlBuilder.cs

```diff
@@ -56,40 +55,217 @@ public String BuildUrl(String path)
+            Dictionary<String, Boolean> proxyStatus = CheckProxyStatus(path);
+            Object pathIsProxy = proxyStatus["isProxy"];
+            Object proxyIsEncoded = proxyStatus["isEncoded"];
+
+            if (pathIsProxy.Equals(true) && proxyIsEncoded.Equals(false))
+            {
+                // Use WebUtility.UrlEncode to ensure *all* characters are handled,
+                // since it's being used as a path.
+                return WebUtility.UrlEncode(path);
+            }
+            else if (pathIsProxy.Equals(true) && proxyIsEncoded.Equals(true))
+            {
+                // Do nothing to URL being used as path since its encoded.
+                return path;
+            }
+            else
+            {
+                // The path is not being used as a proxy, so leave
+                // legal characters like '/' and '@' alone.
+                return EncodeURI(path);
+            }
+        }
+
+        /*
+        * EncodeURI encodes string `p` to the UTF8 format. It splits
+        * the string on the `/` character to avoid accidentally
+        * encoding them. It then iterates over every character and
+        * calls the `UrlEncode` method on them. Finally, it joins
+        * the string array on `/`, adding the removed character back.
+        */
+        private String EncodeURI(String p)
+        {
+            string[] splitPath = p.Split('/');
+            List<String> sanitizedPath = new List<String>();
+
+            foreach (String s in splitPath)
+            {
+                sanitizedPath.Add(UrlEncode(s));
+            }
+
+            return string.Join("/", sanitizedPath);
+        }
+
+
+        /* 
+        * CheckProxyStatus checks if the path has one of the four possible
+        * acceptable proxy prefixes. First we check if the path has the
+        * correct ascii prefix. If it does then we know that it is a proxy,
+        * but it's not percent encoded. Second, we check if the path is
+        * prefixed by a percent-encoded prefix. If it is, we know that it's
+        * a proxy and that it's percent-encoded. Finally, if the path isn't
+        * prefixed by any of these four prefixes, it is not a valid proxy.
+        * This might be "just enough validation," but if we run into issues
+        * we can make this check smarter/more-robust.
+        */
+        private static Dictionary<String, Boolean> CheckProxyStatus(String p)
+        {
+            String path = p;
+            Dictionary<String, Boolean> status = new Dictionary<String, Boolean>();
+            path.Replace("^/", "");
+
+            String asciiHTTP = "http://";
+            String asciiHTTPS = "https://";
+
+            String encodedHTTP = "http%3A%2F%2F";
+            String encodedHTTPS = "https%3A%2F%2F";
 
-            if (path.StartsWith("http"))
+            String encodedHTTPLower = "http%3a%2f%2f";
+            String encodedHTTPSLower = "https%3a%ff%2f";
```

Comment:
We have to leave pre-encoded path support in since the library has, in the past, supported it. As evidenced by `UrlBuilderProperlySignsFullyQualifiedUrls` `UrlBuilderProperlySignsFullyQualifiedUrlsWithParameters` tests.